### PR TITLE
Cleaning device net system

### PR DESCRIPTION
**Pull request by wrexbe**
I think I made it faster. This system feels flawed, need to think about it more.

### CODE REVIEW

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs [hunk 1]
```diff
@@ -219,70 +190,6 @@ private void SendToConnections(HashSet<DeviceNetworkComponent> connections, Devi
...
-                return;
-
-            if (!_networks.TryGetValue(device.DeviceNetId, out var deviceNet) || !deviceNet.RandomizeAddress(device.Address, address))
-            {
-                var prefix = string.IsNullOrWhiteSpace(device.Prefix) ? null : Loc.GetString(device.Prefix);
-                device.Address = $"{prefix}{_random.Next():x}";
-                device.CustomAddress = false;
-            }
-        }
-        #endregion
```
Comment (juliangiebel) [inline Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:285]: Why did you remove those methods?

**Review by juliangiebel — CHANGES_REQUESTED** [review]
Changes are good.
You just removed the regions and all of the component setter methods.

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs (hunk 1 above)
Comment (ElectroJr) [reply Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:285]: I'm guessing the reason was just because currently they're unused.

But they will be useful whenever devices are added that can actually have their address or frequencies set by users, as opposed to the current devices which just use random addresses on pre determined frequencies. So the functions should probably stay.

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs [hunk 2]
```diff
@@ -125,82 +124,54 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
...
+                var totalDevices = 0;
+                if (network.ReceiveAllDevices.TryGetValue(packet.Frequency, out var devices))
+                {
+                    totalDevices += devices.Count;
+                }
                 if (TryGetDevice(packet.NetId, packet.Address, out var device) && device.ReceiveFrequency == packet.Frequency)
-                    recipients.Add(device);
+                {
+                    totalDevices += 1;
+                }
```
Comment (ElectroJr) [inline Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:159]: now that its not a hashset you have to check whether the intended recipient has receive all set to true, because otherwise it gets included twice

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs [hunk 3]
```diff
@@ -125,82 +124,54 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
...
+                var deviceCopy = ArrayPool<DeviceNetworkComponent>.Shared.Rent(totalDevices);
+                if(devices != null)
+                {
+                    devices.CopyTo(deviceCopy);
+                }
+                if(device != null)
+                {
+                    deviceCopy[totalDevices - 1] = device;
+                }
+                SendToConnections(deviceCopy, packet);
```
Comment (ElectroJr) [inline Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:169]: So I'm not very familiar with array pools, but my understanding is that deviceCopy will be some array with a length of at least totalDevices, but may be longer & have devices from some older rented array. So AFAIK you need to pass totalDevices to SendToConnections and ensure that it only iterates over the elements of deviceCopy that it should actually be sending things to?

Comment (wrexbe) [reply Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs]: Not sure if it can have old ones either, but the size is now handled

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs (hunk 2 above)
Comment (wrexbe) [reply Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs]: It is checked now

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs (hunk 1 above)
Comment (wrexbe) [reply Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:285]: I don't trust unused code. I added equivalent functions to address your concerns.

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs [hunk 4]
```diff
@@ -125,82 +128,110 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
...
+                if (TryGetDevice(packet.NetId, packet.Address, out var device) && !device.ReceiveAll && device.ReceiveFrequency == packet.Frequency)
+                {
+                    totalDevices += 1;
+                }
+                var deviceCopy = ArrayPool<DeviceNetworkComponent>.Shared.Rent(totalDevices);
+                if(devices != null)
+                {
+                    devices.CopyTo(deviceCopy);
+                }
+                if(device != null)
```
Comment (ElectroJr) [inline Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:225]:
```suggestion
                if(device != null && !device.ReceiveAll)
```

File: Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs [hunk 5]
```diff
@@ -125,82 +128,110 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
...
-            return new();
+            var deviceNet = GetNetwork(device.DeviceNetId);
+            deviceNet.Remove(device);
+            device.CustomAddress = true;
+            device.Address = address;
+            deviceNet.Add(device);
+        }
+
+        public void RandomizeAddress(EntityUid uid, DeviceNetworkComponent? device = null)
+        {
```
Comment (ElectroJr) [inline Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs:181]: A device will use the same random address when connecting or disconnecting from a network even if it was randomly selected. It only randomizes it if it doesn't have one or the old one is invalid. So this function doesn't actually randomize the address. So it either needs updating or a different name.

**Review by ElectroJr — APPROVED** [review]
